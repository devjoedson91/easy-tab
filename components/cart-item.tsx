import { Card, CardContent } from "@/components/ui/card";
import { formatPrice } from "@/lib/utils";
import { useCartReducer } from "@/store/reducers/cartReducer/useCartReducer";
import { Product } from "@prisma/client";
import { Minus, Plus } from "lucide-react";
import Image from "next/image";

export interface CartProps extends Product {
  amount: number;
  total: number;
}

interface CartItemProps {
  item: CartProps;
}

export default function CartItem({ item }: CartItemProps) {
  const { addToCart, removeFromCart } = useCartReducer();

  function handleRemoveItemCart() {
    removeFromCart(item.id);
  }

  function handleAddItemCart() {
    addToCart(item);
  }

  return (
    <Card className="bg-secondary border-none">
      <CardContent className="py-3 px-4 w-full flex items-center justify-between">
        <div className="relative h-32 w-32">
          <Image
            src={item.banner}
            alt={item.name}
            fill
            className="rounded-md object-cover"
          />
        </div>

        <div className="flex flex-col w-36 gap-4">
          <div className="flex flex-col gap-4">
            <h1 className="text-base text-white font-medium">{item.name}</h1>
            <p className="text-yelowDescription text-base font-medium">
              {formatPrice(item.total)}
            </p>
          </div>
          <div className="border p-2 border-zinc800 rounded-md flex items-center justify-between">
            <button
              onClick={handleRemoveItemCart}
              disabled={(item.amount as number) === 0 || !item}
              data-testid="remove-item"
            >
              <Minus size={22} color="#979797" />
            </button>
            <h1 className="font-bold text-base text-yelowDescription">
              {item.amount}
            </h1>
            <button onClick={handleAddItemCart}>
              <Plus size={22} color="#979797" />
            </button>
          </div>
        </div>
      </CardContent>
    </Card>
  );
}
