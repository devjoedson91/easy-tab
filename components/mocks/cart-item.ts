import { CartProps } from "../cart-item";

export const cartPropsMock: CartProps = {
  amount: 1,
  banner: "https://utfs.io/f/b703fcaa-eb9c-4257-a08e-fba0f0e12fc1-pr8gxl.png",
  created_at: new Date(),
  category_id: "092fc284-449e-41c1-ae5f-ff9beb88b358",
  description: "Lorem ipsum dolor sit amet",
  id: "507ac986-e266-4081-b1e3-c2d142dc8b70",
  name: "Sorvete Especial",
  price: "30",
  restaurantId: "6a0fb54d-c4d0-40db-841a-2dee2f7f2ee7",
  total: 30,
  updated_at: new Date(),
};
