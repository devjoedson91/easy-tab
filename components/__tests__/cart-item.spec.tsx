import { screen, render, fireEvent } from "@testing-library/react";
import CartItem, { CartProps } from "../cart-item";
import { Provider } from "react-redux";
import store from "@/store";
import { cartPropsMock } from "../mocks/cart-item";

const renderComponent = (cart: CartProps) => {
  return render(
    <Provider store={store}>
      <CartItem item={cart} />
    </Provider>
  );
};

describe("<CartItem />", () => {
  it("should rendered correctly", () => {
    const { getByRole, getByText } = renderComponent(cartPropsMock);

    const productName = getByText(/sorvete especial/i);

    expect(getByRole("img", { name: cartPropsMock.name })).toBeInTheDocument();

    expect(
      getByRole("heading", { name: cartPropsMock.amount.toString() })
    ).toBeInTheDocument();

    expect(productName).toBeInTheDocument();
  });

  it("should format the number as BRL currency", () => {
    const { getByText } = renderComponent(cartPropsMock);

    expect(getByText("R$ 30,00")).toBeInTheDocument();
  });

  it("it should match snapshot", () => {
    const { container } = renderComponent(cartPropsMock);

    expect(container.firstChild).toMatchSnapshot();
  });

  it("items amount control", () => {
    const { getByTestId } = renderComponent(cartPropsMock);

    const button = getByTestId("remove-item");

    fireEvent.click(button);
  });
});
