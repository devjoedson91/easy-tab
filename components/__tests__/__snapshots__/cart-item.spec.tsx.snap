// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<CartItem /> it should match snapshot 1`] = `
<div
  class="rounded-lg border border-slate-200 text-slate-950 shadow-sm dark:border-slate-800 dark:bg-slate-950 dark:text-slate-50 bg-secondary border-none"
>
  <div
    class="p-6 py-3 px-4 w-full flex items-center justify-between"
  >
    <div
      class="relative h-32 w-32"
    >
      <img
        alt="Sorvete Especial"
        class="rounded-md object-cover"
        data-nimg="fill"
        decoding="async"
        loading="lazy"
        sizes="100vw"
        src="/_next/image?url=https%3A%2F%2Futfs.io%2Ff%2Fb703fcaa-eb9c-4257-a08e-fba0f0e12fc1-pr8gxl.png&w=3840&q=75"
        srcset="/_next/image?url=https%3A%2F%2Futfs.io%2Ff%2Fb703fcaa-eb9c-4257-a08e-fba0f0e12fc1-pr8gxl.png&w=640&q=75 640w, /_next/image?url=https%3A%2F%2Futfs.io%2Ff%2Fb703fcaa-eb9c-4257-a08e-fba0f0e12fc1-pr8gxl.png&w=750&q=75 750w, /_next/image?url=https%3A%2F%2Futfs.io%2Ff%2Fb703fcaa-eb9c-4257-a08e-fba0f0e12fc1-pr8gxl.png&w=828&q=75 828w, /_next/image?url=https%3A%2F%2Futfs.io%2Ff%2Fb703fcaa-eb9c-4257-a08e-fba0f0e12fc1-pr8gxl.png&w=1080&q=75 1080w, /_next/image?url=https%3A%2F%2Futfs.io%2Ff%2Fb703fcaa-eb9c-4257-a08e-fba0f0e12fc1-pr8gxl.png&w=1200&q=75 1200w, /_next/image?url=https%3A%2F%2Futfs.io%2Ff%2Fb703fcaa-eb9c-4257-a08e-fba0f0e12fc1-pr8gxl.png&w=1920&q=75 1920w, /_next/image?url=https%3A%2F%2Futfs.io%2Ff%2Fb703fcaa-eb9c-4257-a08e-fba0f0e12fc1-pr8gxl.png&w=2048&q=75 2048w, /_next/image?url=https%3A%2F%2Futfs.io%2Ff%2Fb703fcaa-eb9c-4257-a08e-fba0f0e12fc1-pr8gxl.png&w=3840&q=75 3840w"
        style="position: absolute; height: 100%; width: 100%; left: 0px; top: 0px; right: 0px; bottom: 0px; color: transparent;"
      />
    </div>
    <div
      class="flex flex-col w-36 gap-4"
    >
      <div
        class="flex flex-col gap-4"
      >
        <h1
          class="text-base text-white font-medium"
        >
          Sorvete Especial
        </h1>
        <p
          class="text-yelowDescription text-base font-medium"
        >
          R$Â 30,00
        </p>
      </div>
      <div
        class="border p-2 border-zinc800 rounded-md flex items-center justify-between"
      >
        <button
          data-testid="remove-item"
        >
          <svg
            class="lucide lucide-minus"
            fill="none"
            height="22"
            stroke="#979797"
            stroke-linecap="round"
            stroke-linejoin="round"
            stroke-width="2"
            viewBox="0 0 24 24"
            width="22"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M5 12h14"
            />
          </svg>
        </button>
        <h1
          class="font-bold text-base text-yelowDescription"
        >
          1
        </h1>
        <button>
          <svg
            class="lucide lucide-plus"
            fill="none"
            height="22"
            stroke="#979797"
            stroke-linecap="round"
            stroke-linejoin="round"
            stroke-width="2"
            viewBox="0 0 24 24"
            width="22"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M5 12h14"
            />
            <path
              d="M12 5v14"
            />
          </svg>
        </button>
      </div>
    </div>
  </div>
</div>
`;
